Solaris Fixes

HeinrichHartmann opened the pull request:
On operating systems like Solaris and Illumos where the stack grows down, lightuserdata and user data cannot be assigned from the stack. This commit patch fixes this behaviour by moving the allocations to the heap (using calloc).

We also ran into symbol collisions (`compile`) when linking the library with our codebase. This is also addressed in a separate commit.

HeinrichHartmann committed:
Upgrade to Version 1.0

as provided by http://www.inf.puc-rio.br/~roberto/lpeg/ on 2016-05-26

`MD5 (lpeg-1.0.0.tar.gz) = 0aec64ccd13996202ad0c099e2877ece`

HeinrichHartmann committed:
Allocate lightuserdata on the heap.

Solaris (OmniOS, Illumos) version of luajit does not support support
allocation of lightuserdata on the stack. This gives errors like this:

```
$ luajit test.lua
General tests for LPeg library
version 1.0.0
luajit: bad light userdata pointer
stack traceback:
        [C]: in function 'match'
        test.lua:58: in main chunk
        [C]: at 0x00402d50
```

This commit patch fixes this behavior by moving the allocations to the
heap.

HeinrichHartmann committed:
Enforce usage of our compile() implementation by renaming it

Original Author: Jonas Kunze.

harningt commented on lpvm.c:
```diff
@@ -146,7 +147,7 @@ static int removedyncap (lua_State *L, Capture *capture,
 const char *match (lua_State *L, const char *o, const char *s, const char *e,
                    Instruction *op, Capture *capture, int ptop) {
-  Stack stackbase[INITBACK];
+  Stack *stackbase = calloc(INITBACK, sizeof(Stack));
```
No sanity check present in case of failure.

harningt commented on lptree.c:
```diff
@@ -1147,9 +1148,10 @@ static size_t initposition (lua_State *L, size_t len) {
 */
 static int lp_match (lua_State *L) {
-  Capture capture[INITCAPSIZE];
+  Capture *capture = calloc(INITCAPSIZE, sizeof(Capture));
```
No sanity check in case of failure.

HeinrichHartmann committed:
Add checks for calloc's return code.

By using the assert macro the process will exit if calloc fails.  There
is little chance to recover from a failed allocation. Moreover assert()
is used at other places already.
